### PR TITLE
[Merged by Bors] - feat: algebraic versions of LocallyConstant.comap

### PR DESCRIPTION
**Pull request by dagurtomas**
We give algebraic versions of `LocallyConstant.comap` given the relevant algebraic structure on the target. 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/LocallyConstant/Algebra.lean [hunk 1]
```diff
@@ -291,4 +291,65 @@ theorem coe_algebraMap (r : R) : ⇑(algebraMap R (LocallyConstant X Y) r) = alg
...
+    LocallyConstant Y Z →* LocallyConstant X Z where
+  toFun := comap f
+  map_one' := by
+    ext x
+    rw [coe_comap_apply _ _ hf]
+    rfl
+  map_mul' := by
+    intro r s
+    ext x
+    simp
```
Comment (eric-wieser) [inline Mathlib/Topology/LocallyConstant/Algebra.lean:310]: This is a non-terminal simp, which should be squeezed

File: Mathlib/Topology/LocallyConstant/Algebra.lean [hunk 2]
```diff
@@ -291,4 +291,65 @@ theorem coe_algebraMap (r : R) : ⇑(algebraMap R (LocallyConstant X Y) r) = alg
...
+noncomputable
+def comapMul  [MulOneClass Z] (f : X → Y) (hf : Continuous f) :
+    LocallyConstant Y Z →* LocallyConstant X Z where
+  toFun := comap f
+  map_one' := by
+    ext x
+    rw [coe_comap_apply _ _ hf]
+    rfl
+  map_mul' := by
+    intro r s
```
Comment (eric-wieser) [inline Mathlib/Topology/LocallyConstant/Algebra.lean:308]:
```suggestion
  map_mul' r s := by
```

File: Mathlib/Topology/LocallyConstant/Algebra.lean [hunk 3]
```diff
@@ -291,4 +291,63 @@ theorem coe_algebraMap (r : R) : ⇑(algebraMap R (LocallyConstant X Y) r) = alg
...
+    rfl
+  map_mul' r s := by
+    ext x
+    dsimp only [coe_mul, Pi.mul_apply]
+    rw [coe_comap_apply _ _ hf, coe_comap_apply _ _ hf, coe_comap_apply _ _ hf]
+    rfl
+
+/-- `LocallyConstant.comap` is a linear map. -/
+noncomputable
+def comapLinear {R : Type _} [Semiring R] [AddCommMonoid Z] [Module R Z] (f : X → Y)
```
Comment (ADedecker) [inline Mathlib/Topology/LocallyConstant/Algebra.lean:315]:
```suggestion
def comapLinearMap {R : Type _} [Semiring R] [AddCommMonoid Z] [Module R Z] (f : X → Y)
```

Another possible name for brevity would be `comapₗ`

File: Mathlib/Topology/LocallyConstant/Algebra.lean [hunk 4]
```diff
@@ -291,4 +291,63 @@ theorem coe_algebraMap (r : R) : ⇑(algebraMap R (LocallyConstant X Y) r) = alg
...
+    ext x
+    dsimp only [RingHom.id_apply, coe_add, Pi.add_apply, AddHom.toFun_eq_coe,
+      AddHom.coe_mk, coe_smul, Pi.smul_apply]
+    rw [coe_comap_apply _ _ hf, coe_comap_apply _ _ hf]
+    rfl
+
+lemma comapLinear_injective {R : Type _} [Semiring R] [AddCommMonoid Z] [Module R Z] (f : X → Y)
+    (hf : Continuous f) (hfs : Function.Surjective f) :
+    LinearMap.ker (comapLinear f hf : LocallyConstant Y Z →ₗ[R] LocallyConstant X Z) = ⊥ :=
+  LinearMap.ker_eq_bot_of_injective <| comap_injective _ hf hfs
```
Comment (ADedecker) [inline Mathlib/Topology/LocallyConstant/Algebra.lean:334]: Usually when we put `injective` we actually use `Function.injective`, the kernel result should be called `ker_comapLinear`. See e.g [LinearMap.lsmul_injective](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/BilinearMap.html#LinearMap.lsmul_injective) vs [LinearMap.ker_lsmul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/BilinearMap.html#LinearMap.ker_lsmul)

File: Mathlib/Topology/LocallyConstant/Algebra.lean [hunk 5]
```diff
@@ -291,4 +291,63 @@ theorem coe_algebraMap (r : R) : ⇑(algebraMap R (LocallyConstant X Y) r) = alg
...
+def equivLinear {R : Type _} [Semiring R] [AddCommMonoid Z] [Module R Z] (e : X ≃ₜ Y) :
+    LocallyConstant X Z ≃ₗ[R] LocallyConstant Y Z where
+  toFun := (equiv e).toFun
+  map_smul' := (comapLinear _ e.continuous_invFun).map_smul'
+  map_add' r s := by -- note: (comapLinear e.symm hf).map_add' r s -- doesn't work.
+    ext x
+    dsimp [equiv]
+    have hf : Continuous ↑(e.symm) := e.continuous_invFun
+    rw [coe_comap_apply _ _ hf, coe_comap_apply _ _ hf, coe_comap_apply _ _ hf]
+    rfl
```
Comment (ADedecker) [inline Mathlib/Topology/LocallyConstant/Algebra.lean:346]:
```suggestion
  map_add' := (comapAddMonoidHom _ e.continuous_invFun).map_add'
```

The reason it didn't work is that, contrary to the `map_smul'` field, there's no way to infer the ring `R` here. You could also do:
```lean
  map_add' := (comapLinear (R := R) _ e.continuous_invFun).map_add'
```
But it's cleaner to just use the additive version.

File: Mathlib/Topology/LocallyConstant/Algebra.lean [hunk 6]
```diff
@@ -291,4 +291,63 @@ theorem coe_algebraMap (r : R) : ⇑(algebraMap R (LocallyConstant X Y) r) = alg
...
+    rw [coe_comap_apply _ _ hf, coe_comap_apply _ _ hf]
+    rfl
+
+lemma comapLinear_injective {R : Type _} [Semiring R] [AddCommMonoid Z] [Module R Z] (f : X → Y)
+    (hf : Continuous f) (hfs : Function.Surjective f) :
+    LinearMap.ker (comapLinear f hf : LocallyConstant Y Z →ₗ[R] LocallyConstant X Z) = ⊥ :=
+  LinearMap.ker_eq_bot_of_injective <| comap_injective _ hf hfs
+
+noncomputable
+def equivLinear {R : Type _} [Semiring R] [AddCommMonoid Z] [Module R Z] (e : X ≃ₜ Y) :
```
Comment (ADedecker) [inline Mathlib/Topology/LocallyConstant/Algebra.lean:337]:
```suggestion
def congrLeftLinearEquiv {R : Type _} [Semiring R] [AddCommMonoid Z] [Module R Z] (e : X ≃ₜ Y) :
```

Or `congrLeftₗ`

**Comment by ADedecker** [comment]
Could you merge master?

Also, while you're at it, could you do the following cleanup in `Topology.LocallyCostant.Basic`?
- change [LocallyConstant.coe_comap](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/LocallyConstant/Basic.html#LocallyConstant.coe_comap) to
```lean
@[simp]
theorem coe_comap (f : X → Y) (g : LocallyConstant Y Z) (hf : Continuous f) :
    comap f g = g ∘ f := by
  rw [comap, dif_pos hf]
  rfl
```
- change [LocallyConstant.comap_id](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/LocallyConstant/Basic.html#LocallyConstant.comap_id) to
```lean
theorem comap_id : comap (@id X) = @id (LocallyConstant X Z) := by
  ext
  simp only [continuous_id, id.def, Function.comp.right_id, coe_comap]
```
- change [LocallyConstant.comap_const](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/LocallyConstant/Basic.html#LocallyConstant.comap_const) to
```lean
theorem comap_const (f : X → Y) (y : Y) (h : ∀ x, f x = y) :
    (comap f : LocallyConstant Y Z → LocallyConstant X Z) = fun g => (const X (g y)) := by
  ext; rw [coe_comap]
  · simp only [Function.comp_apply, h, coe_const, Function.const_apply]
  · rw [show f = fun _ => y by ext; apply h]
    exact continuous_const
```

Thanks!

**Comment by dagurtomas** [comment]
Thanks for the helpful review!

**Pull request closed without merging**